dams/src/audit_log.rs

```
@@ -0,0 +1,62 @@
+//! Audit log entries, fields, and types
+//!
+//! Includes possible actions to log and outcomes of those actions
+
+use crate::user::UserId;
+
+use mongodb::bson::oid::ObjectId;
+use mongodb::bson::DateTime;
+use serde::{Deserialize, Serialize};
+use std::fmt::{Display, Formatter};
+
+/// Options for which action was taken for a given [`LogEntry`]
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Action {
+    Register,
+    Login,
+}
+
+/// Options for the outcome of a given action in a [`LogEntry`]
+#[derive(Debug, Serialize, Deserialize)]
+pub enum Outcome {
+    Successful,
+    Failed,
+}
```

I think this is good for now but we may want to tie these to the actions in the future. I can see this type becoming very large if there are a range of different outcomes for different actions. It would be easy to make mistakes if that ends up happening.

Just wanted to make a note of this. As long as we keep a consistent format for the database representation of each entry, this should be easy to refactor down the line. For now, I think what you've got is the best way to do this.

---

Yeah, I think you're right that we will eventually need to refactor the outcome-action setup, once we know what we can report back to the user.  For context, I think we still haven't decided the level of granularity we can return for the outcome of certain actions (e.g. for OPAUQE processes, I believe all types of failures need to appear the same to the user). @marsella, does that sounds right?

---

Yeah, for authentication we only want to report a single failure type. You're both correct that we will want to reveal more information for future actions (e.g. if the policy engine rejects a signature later, we want to know which asset fiduciaries rejected it, and maybe why). But without those behaviors spec'd out, I don't know whether it would be best to add a context string to the existing `Failure` vs add more variants to express wider range of outcomes vs some other option.